lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -34,421 +25,546 @@
+
+      // Make the first port as the base value.
+      // Note: this is a port and the actual reference base exists in another
+      // module.
+      auto leader = rwProbeClasses.getLeaderValue(sinkNode);
+      rwProbeRefersTo[leader] = basePortNode;
+
+      setOfEquivalentRWProbes.insert(sinkNode);
+      // Add a reaching def  from base RWProbe port to all other RWProbe ports.
+      for (auto probe : setOfEquivalentRWProbes)
+        if (probe != basePortNode)
+          reachingDefGraph[probe].second.push_back(basePortNode);
+    }
+  }
+
+  // Record the FieldRef, corresponding to the result of the sub op
+  // `result = base[index]`
+  void recordValueRefersToFieldRef(Value base, unsigned fieldID, Value result) {
+
+    // Check if base is itself field of an aggregate.
+    auto it = valToFieldRefs.find(base);
+    if (it != valToFieldRefs.end()) {
+      // Rebase it to the original aggregate.
+      // Because of subaccess op, each value can refer to multiple FieldRefs.
+      SmallVector<FieldRef> entry;
+      for (auto &sub : it->second)
+        entry.emplace_back(sub.getValue(), sub.getFieldID() + fieldID);
+      // Update the map at the end, to avoid invaliding the iterator.
+      valToFieldRefs[result].append(entry.begin(), entry.end());
+      return;
+    }
+    // Break cycles from registers.
+    if (auto *def = base.getDefiningOp()) {
+      if (isa<RegOp, RegResetOp, SubfieldOp, SubindexOp, SubaccessOp>(def))
         return;
-      auto memPort = ref.getValue();
-      auto type = type_cast<BundleType>(memPort.getType());
-      auto enableFieldId = type.getFieldID((unsigned)ReadPortSubfield::en);
-      auto dataFieldId = type.getFieldID((unsigned)ReadPortSubfield::data);
-      auto addressFieldId = type.getFieldID((unsigned)ReadPortSubfield::addr);
-      if (ref.getFieldID() == enableFieldId ||
-          ref.getFieldID() == addressFieldId) {
-        for (auto dataField : fieldToVals[FieldRef(memPort, dataFieldId)])
-          children.push_back(dataField);
-      }
     }
+    valToFieldRefs[result].emplace_back(base, fieldID);
   }
 
-  void reportLoopFound(Value childVal, VisitingSet visiting) {
-    // TODO: Work harder to provide best information possible to user,
-    // especially across instances or when we trace through aliasing values.
-    // We're about to exit, and can afford to do some slower work here.
-    auto getName = [&](Value v) {
-      if (isa_and_nonnull<SubfieldOp, SubindexOp, SubaccessOp>(
-              v.getDefiningOp())) {
-        assert(!valRefersTo[v].empty());
-        // Pick representative of the "alias set", not deterministic.
-        return getFieldName(*valRefersTo[v].begin()).first;
+  void handleMemory(MemOp mem) {
+    if (mem.getReadLatency() > 0)
+      return;
+    // Add the enable and address fields as the reaching definitions to the data
+    // field.
+    for (auto memPort : mem.getResults())
+      // TODO: Can reftype ports create cycle ?
```

Comment:
You mean cycles involving debug ports on memories? Seems low priority but I would imagine they might.